### ISSUE TITLE
Support for clients-related commands?

### ISSUE THREAD

**Issue by hjri** [janakj/pulseaudio.js#2]
I'm making a tool that manipulates other client streams, i.e. move active window's stream to a different device via a MIDI controller, this requires access to list of clients (`GET_CLIENT_INFO_LIST` aka `pactl list clients`) and ability to move clients to different sinks (`MOVE_SINK_INPUT`, `MOVE_SOURCE_OUTPUT` - `pactl move-sink-input` and `move-source-output` respectively), ability to set clients volume would also be welcome.

**Comment by janakj** [comment]
These commands are simple to implement if you know what parameters the server expects and what it returns in the response. That may require looking at the official PulseAudio client library because the low-level protocol is not very well document.

Contributions are welcome and I am happy to help you get the new commands implemented if you are willing to put some effort into it.

**Comment by hjri** [comment]
yeah, i'm currently implementing them in my project locally by importing from `node_modules/pulseaudio.js/dist` (need to extend `Command` which isn't exported by default I believe).

I currently implemented simple thingy for the list in JS and working on the `moveSinkInput` command:
```js
export async function getClients (pa) {
  class GetClientsList extends Command {
    processResponse(packet) {
      const rv = [];
      while (packet.i < packet.body.length) {
        rv.push({
          index: packet.getUInt32(),
          name: packet.getString(),
          owner_module: packet.getUInt32(),
          driver: packet.getString(),
          pa_proplist: packet.getProps()
        });
      }
      return rv;
    }
  }
  const cmd = new GetClientsList(PA_COMMAND.GET_CLIENT_INFO_LIST)

  return cmd.processResponse(await pa._invoke(cmd))
}
```

I will try to make a proper contribution later, but I'm sorta burned out by TypeScript an don't want to deal with more than absolute minimum.

**Comment by janakj** [comment]
Is the above code working? If yes, I can merge a variant of this into the source tree.

Regarding other commands, if you could lookup the parameters and return values, I can do the TypeScript API. All I really need is for somebody to reverse-engineer the official PA client and then to test the result.

**Comment by hjri** [comment]
yeah, the code above is working, but turns out i needed `GetSinkInputList` instead, haha.

I'm currently looking into pulse's source to get moving working

**Comment by hjri** [comment]
code for `sinkInputList`, it contains copypaste from `getSinkInfo` or similar

```js
export async function getSinkInputList (pa) {
  class GetSinkInputList extends Command {
    processResponse(packet) {
      const rv = [];
      while (packet.i < packet.body.length) {
        rv.push({
          // copypaste
          index: packet.getUInt32(),
          name: packet.getString(),
          module: packet.getUInt32(),
          client: packet.getUInt32(),
          sink: packet.getUInt32(),
          sampleSpec: packet.getSampleSpec(),
          channelMap: packet.getChannelMap(),
          volume: packet.getCvolume(),
          latency: {
            minimum: packet.getUsec(),
            maximum: packet.getUsec()
          },
          resampleMethod: packet.getString(),
          driver: packet.getString(),
          muted: packet.getBool(),
          properties: packet.getProps(),
          corked: packet.getBool(),
          hasVolume: packet.getBool(),
          writableVolume: packet.getBool(),
          formatInfo: packet.getFormatInfo()
          // end copypaste
        });
      }
      return rv;
    }
  }
  const cmd = new GetSinkInputList(PA_COMMAND.GET_SINK_INPUT_INFO_LIST)

  return cmd.processResponse(await pa._invoke(cmd))
}
```
this also works

**Comment by hjri** [comment]
Now about the moving part
```js
export async function moveSinkInput (pa, index, sink) {
  class MoveSinkInput extends SelectByIndex {
    constructor(index, sink) {
      super(PA_COMMAND.MOVE_SINK_INPUT, index);
      switch (typeof sink) {
        case 'number':
          this.addUInt32(sink);
          this.addString(null);
          break;
        case 'string':
          this.addUInt32(0xffffffff); // PA_INVALID_INDEX
          this.addString(sink);
          break;
      }
    };
  }
  const cmd = new MoveSinkInput(index, sink)

  return cmd.processResponse(await pa._invoke(cmd))
}
```

I was stuck for a few hours because error reporting was kinda confusing, and also i thought PA expects `0xfffffffe` which was used somewhere else but it actually expects `0xffffffff`

i used https://github.com/pulseaudio/pulseaudio/blob/master/src/pulse/introspect.c#L2007 as reference basically.

also all my code snippets are WTFPL, just to clear that out.

**Comment by janakj** [comment]
> Now about the moving part
> 
> ```js
> export async function moveSinkInput (pa, index, sink) {
>   class MoveSinkInput extends SelectByIndex {
>     constructor(index, sink) {
>       super(PA_COMMAND.MOVE_SINK_INPUT, index);
>       switch (typeof sink) {
>         case 'number':
>           this.addUInt32(sink);
>           this.addString(null);
>           break;
>         case 'string':
>           this.addUInt32(0xffffffff); // PA_INVALID_INDEX
>           this.addString(sink);
>           break;
>       }
>     };
>   }
>   const cmd = new MoveSinkInput(index, sink)
> 
>   return cmd.processResponse(await pa._invoke(cmd))
> }
> ```
> 
> I was stuck for a few hours because error reporting was kinda confusing, and also i thought PA expects `0xfffffffe` which was used somewhere else but it actually expects `0xffffffff`

0xffffffff means "no index" (the value is string). There's a constant for this called PA_NO_INDEX. The value 0xfffffffe is the maximum index number that any PA resource can have. This is essentially just used to check bounds. So, your use of 0xffffffff here is correct.

**Comment by janakj** [comment]
Thanks a lot for your contribution! I integrated all three functions into the library. You can find the new commits on the branch [client](https://github.com/janakj/pulseaudio.js/tree/client).

I made minor modifications to some of your code to make it consistent with the rest of the library. For example, I renamed some of the properties returned by GetClientsList to use camel case and I changed the name of the props attribute to "properties". You may need to to update your code if you access these.

Would you want to try it out? If it works for you, I am ready to release the version 1.2.0 to Github and the npm registry.

**Comment by hjri** [comment]
Thanks. Other similar functions (like for SourceOutputs shouldn't be too different)

how do i try it out exactly, i don't think i can point npm/yarn to git repository since it needs to be compiled. I just use plain old index.js + node_modules installed via yarn, no build process whatsoever.

**Comment by janakj** [comment]
You can install it directly from github with npm as follows:
```
npm install janakj/pulseaudio.js#client
```
This will automatically compile the package.

**Comment by hjri** [comment]
ok, i'll look into it tomorrow-ish while cleaning up my own code

**Comment by janakj** [comment]
Great, let me know how it goes.

**Comment by hjri** [comment]
yeah that seem to be working, thanks

**Comment by janakj** [comment]
Thanks for letting me know. I will release pulseaudio.js 1.2.0 later today.

**Comment by janakj** [comment]
I just published version 1.2.0 that includes the new methods to the NPM registry. Thank you, your contribution is very much appreciated!
